apm-agent-core/src/test/java/co/elastic/apm/agent/bci/IndyBootstrapTest.java

```diff
@@ -0,0 +1,53 @@
+/*-
+ * #%L
+ * Elastic APM Java agent
+ * %%
+ * Copyright (C) 2018 - 2020 Elastic and contributors
+ * %%
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * #L%
+ */
+package co.elastic.apm.agent.bci;
+
+import co.elastic.apm.agent.AbstractInstrumentationTest;
+import net.bytebuddy.dynamic.loading.ClassInjector;
+import org.junit.jupiter.api.Test;
+import org.stagemonitor.util.IOUtils;
+
+import java.io.InputStream;
+import java.util.Collections;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotEquals;
+
+class IndyBootstrapTest extends AbstractInstrumentationTest {
+
+    @Test
+    void testSetJavaBaseModule() throws Throwable {
+        Module javaBaseModule = Class.class.getModule();
+        assertNotEquals(IndyBootstrapTest.class.getModule(), javaBaseModule);
+
+        // In order to test this functionality, IndyBootstrapDispatcherModuleSetter needs to be loaded from the Boot CL
+        InputStream classFileAsStream = IndyBootstrapDispatcherModuleSetter.class.getResourceAsStream("IndyBootstrapDispatcherModuleSetter.class");
+        byte[] bootstrapClass = IOUtils.readToBytes(classFileAsStream);
+        ClassInjector.UsingUnsafe.ofBootLoader().injectRaw(Collections.singletonMap(IndyBootstrapDispatcherModuleSetter.class.getName(), bootstrapClass));
+
+        IndyBootstrap.setJavaBaseModule(IndyBootstrapTest.class);
```

Comment:
[question] Would it be possible to have an exception thrown with an explicit message when this method is called and `IndyBootstrapDispatcherModuleSetter` hasn't been loaded from the right classloader ? If yes, that might help to diagnose any unexpected state in the field.

Reply:
This can't happen - the `setJavaBaseModule` method tries to load it from the bootstrap CL explicitly, so failing to inject it there will result with `ClassNotFoundException`.
As you can see here, I _do_ load it with the test's class loader, but its single method is never invoked on any class that is not loaded correctly.